src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_SIZE = 10;
+
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[INITIAL_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndex(index, size + 1);
+        ensureSize(size + 1);
+        if (index < size) {
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+        }
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null || list.size() == 0) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        elements[index] = value;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        final T value = (T) elements[index];
+        size--;
+        if (index < size) {
+            System.arraycopy(elements, index + 1, elements, index, size - index);
+        }
+        elements[size] = null;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] == element)
+                    || (elements[i] != null && elements[i].equals(element))) {
+                remove(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException("No such element in the list: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void ensureSize(int size) {
+        if (size >= elements.length) {
+            Object[] newArray = new Object[elements.length * 3 / 2];
+            System.arraycopy(elements, 0, newArray, 0, elements.length);
+            elements = newArray;
+        }
+    }
```

Comment:
use constants instead of numbers
```
    private void grow() {
        if (elements.length == size) {
            T[] newArrays = (T[]) new Object[(int) (elements.length * GROW_FACTORY)];
            System.arraycopy(elements, 0, newArrays, 0, elements.length);
            elements = newArrays;
        }
    }
```

Reply:
Agree.

Reply:
It makes sense.
Thanks!